src/material/datepicker/date-range-input-parts.ts

```diff
@@ -284,6 +288,26 @@ export class MatStartDate<D> extends _MatDateRangeInputBase<D> implements CanUpd
     const value = element.value;
     return value.length > 0 ? value : element.placeholder;
   }
+
+  override _onKeydown(event: KeyboardEvent) {
+    const endInput = this._rangeInput._endInput;
+    const element = this._elementRef.nativeElement;
+    const isLtr = this._dir?.value !== 'rtl';
+
+    // If the user hits RIGHT (LTR) when at the end of the input (and no
+    // selection), move the cursor to the start of the end input.
+    if (
+      ((event.keyCode === RIGHT_ARROW && isLtr) || (event.keyCode === LEFT_ARROW && !isLtr)) &&
+      element.selectionStart === element.value.length &&
+      element.selectionEnd === element.value.length
+    ) {
+      event.preventDefault();
+      endInput._elementRef.nativeElement.setSelectionRange(0, 0);
+      endInput.focus();
+    }
+
+    super._onKeydown(event);
```

Comment:
I think that we'd want to put this in an `else` so we don't accidentally trigger multiple actions. Same goes for the other `keydown` listener.